src/cephadm/cephadm

```diff
@@ -3130,24 +3195,45 @@ def list_daemons(detail=True, legacy_dir=None):
+                                # find that container's info
+                                if adjusted_unit_name in parsed_ps_out[k]['Names']:
+                                    # grab container id, image name and start timestamp from corresponding row in ps output
+                                    container_id = parsed_ps_out[k]['ID']
+                                    image_name = parsed_ps_out[k]['Image']
+                                    start_stamp = parsed_ps_out[k]['StartedAt']
+                            # if an image name was found, can use it to find that image's id
+                            if image_name:
+                                for k in range(0, len(parsed_image_list_out)):
+                                    # match found image name with image name in podman image list output
+                                    if image_name in parsed_image_list_out[k]['names'][0]:
+                                        image_id = parsed_image_list_out[k]['id']
+                        elif 'docker' in container_path:
+                            for k in range(0, len(parsed_ps_out)):
+                                # match container name from get_unit_name with name in docker ps command to
+                                # find that container's info
+                                if adjusted_unit_name in parsed_ps_out[k][2]:
+                                    # grab container id, image name and start timestamp from corresponding row in ps output
+                                    container_id = parsed_ps_out[k][0]
+                                    image_name = parsed_ps_out[k][1]
+                                    start_stamp = parsed_ps_out[k][3]
+                            # if an image name was found, can use it to find that image's id
+                            if image_name:
+                                # image names from docker image list command do not include extra
+                                # info sometimes found after a colon. Ex: "latest-master-devel" in
+                                # "docker.io/ceph/daemon-base:latest-master-devel". Adjusted version removes any
+                                # possible info after a colon so strings match
+                                char_index = unit_name.rfind(":");
+                                adjusted_image_name = image_name.split(':')[0]
+                                # match found image name with image name in podman image list output
+                                for k in range(0, len(parsed_image_list_out)):
+                                    if adjusted_image_name in parsed_image_list_out[k][0]:
+                                        image_id = parsed_image_list_out[k][1]
+
+                        # if container id was found, can use it to try and find version
+                        if container_id and image_id:
```

Comment:
can you make those code blocks dedicated functions? list_daemons is already way too huge.